### PR TITLE
Framework: Add package: @wordpress/babel-plugin-import-jsx-pragma

### LINKED ISSUE #6956: Intermittent e2e test failures
The e2e test suite currently fails intermittently. Sometimes the test failure looks like this:

image

I have no idea on the cause of this, but it's getting in the way of other unrelated development.

### PR DESCRIPTION
**Pull request by aduth**
Fixes #6956
Related: #742 (specifically https://github.com/WordPress/gutenberg/pull/742#issuecomment-300603176 )

This pull request seeks to introduce a new package, `@wordpress/babel-plugin-import-jsx-pragma`, which seeks to serve as a compromise between competing objectives of developer experience (not needing explicit import to use JSX) and eliminating reliance on globals (previously the pragma was configured to a `wp.element.createElement` global). The transform effectively guarantees that the required import is made to be in scope wherever JSX is used, and updates our configured `pragma` from `wp.element.createElement` to simply `createElement`.

[See included README.md](https://github.com/WordPress/gutenberg/blob/c846b72d6503916648b669c2fc2704ea124f56c4/packages/babel-plugin-import-jsx-pragma/README.md)
[See demonstration](https://astexplorer.net/#/gist/1c572e34696714efaf0bd26e53601a15/80259c1dd99155881c41ac348edbb7155a0ee685)

```js
// Before:
const MyComponent = () => <div />;

// After:
import React from "react";
const MyComponent = () => <div />;
```

It resolves #6956 where end-to-end tests fail due to a conflict between the change detection tests reloading the page and core logic to [unset the `window.wp` global on page unload](https://github.com/WordPress/wordpress-develop/blob/78f2b0e/src/js/_enqueues/wp/media/models.js#L239-L242), where previously failures could occur when a render happened after the unsetting (as it would attempt to unsuccessfully call the missing `wp.element.createElement`).

**Testing instructions:**

Verify that the editor loads without issue.

Ensure unit tests pass.

### CODE REVIEW

**Comment by aduth** [comment]
The test failure surfaces an interesting problem, which is that in order to build packages, we now need this packaged Babel plugin (chicken and egg problem). We could either (a) publish the package once manually, then reference the published version or (b) update the build process for packages to first build the required `babel-plugin-import-jsx-pragma` standalone.

**Comment by tofumatt** [comment]
> update the build process for packages to first build the required babel-plugin-import-jsx-pragma standalone

Option "B" seems like the better one to me.

File: packages/babel-plugin-import-jsx-pragma/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+Refer to the [Babel Plugins documentation](http://babeljs.io/docs/en/plugins)
+if you don't yet have experience working with Babel plugins.
+
+Include `@wordpress/babel-plugin-import-jsx-pragma` as a plugin in your Babel
+configuration.
+
+It's assumed that you're also using [@babel/transform-react-jsx](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx/).
+Otherwise, you may encounter errors when encountering JSX tokens.
+
+```
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/README.md]: You should mark this as `js` so it'll syntax-highlight nicely on GitHub, eg:

<pre>
```js
// .babelrc.js
module.exports = {
	plugins: [
		'@wordpress/babel-plugin-import-jsx-pragma',
		'@babel/transform-react-jsx',
	],
};
```
</pre>

File: packages/babel-plugin-import-jsx-pragma/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+```bash
+npm install @wordpress/babel-plugin-import-jsx-pragma
+```
+
+## Usage
+
+Refer to the [Babel Plugins documentation](http://babeljs.io/docs/en/plugins)
+if you don't yet have experience working with Babel plugins.
+
+Include `@wordpress/babel-plugin-import-jsx-pragma` as a plugin in your Babel
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/README.md]: This could be reworded to be instructive and less vague, as for any standard setup there'll be errors, and if the setup is non-standard we can assume developer familiarity. How about just:

```markdown
Include `@wordpress/babel-plugin-import-jsx-pragma` (and [@babel/transform-react-jsx](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx/)) as plugins in your Babel configuration.

If you don't include both you will get errors with encountering JSX tokens.
```

File: packages/babel-plugin-import-jsx-pragma/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	plugins: [
+		'@wordpress/babel-plugin-import-jsx-pragma',
+		'@babel/transform-react-jsx',
+	],
+};
+```
+
+## Options
+
+As the `@babel/transform-react-jsx` plugin offers options to customize the
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/README.md]: Do we usually newline our markdown? If you don't it's nicer to diff/edit, I think doing this is annoying when making changes.

File: packages/babel-plugin-import-jsx-pragma/src/index.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			},
+			ImportDeclaration( path, state ) {
+				if ( hasImportedScopeVariable ) {
+					return;
+				}
+
+				const { scopeVariable, isDefault } = getOptions( state.opts );
+
+				// The module source isn't verified, since if at least the
+				// required variable is within scope, its assumed to be
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/src/index.js]: its -> it's

File: packages/babel-plugin-import-jsx-pragma/src/index.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+				hasJSX = true;
+			},
+			ImportDeclaration( path, state ) {
+				if ( hasImportedScopeVariable ) {
+					return;
+				}
+
+				const { scopeVariable, isDefault } = getOptions( state.opts );
+
+				// The module source isn't verified, since if at least the
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/src/index.js]: I think I get this comment but it's a bit hard to parse, especially because "since" could be "because" to make it read more logically.

I don't have ideas for copyediting right now but could this be explained another way? Or maybe broken up into a few sentences so it's easier to parse?

File: packages/babel-plugin-import-jsx-pragma/src/index.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+				// required variable is within scope, its assumed to be
+				// compatible with the targeted transform, and otherwise would
+				// conflict as a duplicate import if introduced separately.
+
```
Comment (tofumatt) [inline packages/babel-plugin-import-jsx-pragma/src/index.js]: This newline is weird to me.

**Review by tofumatt — COMMENTED** [review]
I think this is nice and I just have some thoughts on the docs.

File: packages/babel-plugin-import-jsx-pragma/src/index.js (hunk 6 above)
Comment (aduth) [reply packages/babel-plugin-import-jsx-pragma/src/index.js]: >This newline is weird to me.

For context, there was previously a condition here which checked that the import source matched what we were expecting, but specifically I ran into issues with `@wordpress/element` where we import `createElement` from React proper:

https://github.com/WordPress/gutenberg/blob/04c0ea207efda81d98963c10e09c3c21705212fe/packages/element/src/index.js#L4-L15

I will update the comment and shift it down a line to describe the behavior as part of what it is we're actually testing now.

File: packages/babel-plugin-import-jsx-pragma/README.md (hunk 3 above)
Comment (aduth) [reply packages/babel-plugin-import-jsx-pragma/README.md]: >Do we usually newline our markdown? If you don't it's nicer to diff/edit, I think doing this is annoying when making changes.

We have at some points. This is my first go at it. In retrospect, I agree that I don't like it, and we should probably discourage it. Updated in aad1bed.

**Comment by aduth** [comment]
Thanks for the review @tofumatt . I've pushed a few updates, including one to prebuild the `@wordpress/babel-plugin-import-jsx-pragma` separate from other packages.

File: bin/packages/get-babel-config.js [hunk 7]
```diff
@@ -19,6 +19,14 @@ const plugins = map( babelDefaultConfig.plugins, ( plugin ) => {
 	return plugin;
 } );
 
+if ( process.env.TRANSFORM_JSX_PRAGMA ) {
+	plugins.push( [ require( '../../packages/babel-plugin-import-jsx-pragma' ).default, {
+		scopeVariable: 'createElement',
+		source: '@wordpress/element',
+		isDefault: false,
+	} ] );
+}
```
Comment (youknowriad) [inline bin/packages/get-babel-config.js]: Do you think we should include this in the babel preset package?

Comment (aduth) [reply bin/packages/get-babel-config.js]: Yes 😄 It's been an awkward struggle thus far with the preset. I imagine it will be made easier once it's part of Gutenberg proper.

File: bin/packages/get-babel-config.js [hunk 8]
```diff
@@ -19,6 +19,14 @@ const plugins = map( babelDefaultConfig.plugins, ( plugin ) => {
 	return plugin;
 } );
 
+if ( process.env.TRANSFORM_JSX_PRAGMA ) {
```
Comment (youknowriad) [inline bin/packages/get-babel-config.js]: Also not sure why we need all these env variables (this one and the include/exclude ones)
Can't we just always use this plugin? What's the issue if always include it?

Comment (aduth) [reply bin/packages/get-babel-config.js]: See https://github.com/WordPress/gutenberg/pull/7493#issuecomment-399548925

File: packages/babel-plugin-import-jsx-pragma/src/index.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+			};
+		}
+
+		return _options;
+	}
+
+	return {
+		visitor: {
+			JSXElement() {
+				hasJSX = true;
```
Comment (youknowriad) [inline packages/babel-plugin-import-jsx-pragma/src/index.js]: I read somewhere in babel docs that it's not great to rely on state variables like that and that it's probably better to rely on the "state" object passed to the visitors because of the way these visitors can be called in random orders.

I'm not a babel-plugin developer so I can't argue if it's legit here or not, just wanted to mention it.

**Review by youknowriad — COMMENTED** [review]
LGTM globally, the issue I have is these global variables, it makes it hard to follow. When should I use these globals and when not. I think it's better if it's handled automatically somehow.

File: bin/packages/get-babel-config.js (hunk 8 above)
Comment (youknowriad) [reply bin/packages/get-babel-config.js]: I see so I guess this would solve the issue right? #7556

**Review by youknowriad — APPROVED** [review]

File: packages/babel-plugin-import-jsx-pragma/src/index.js (hunk 9 above)
Comment (aduth) [reply packages/babel-plugin-import-jsx-pragma/src/index.js]: Based on the handbook description of visiting behavior, and particularly with its evaluation on the `Program.exit` pass, I don't know that it should be a concern.

https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#visitors

That said, I'll take a look to see if it's easy enough to switch to using state.

**Comment by aduth** [comment]
>the issue I have is these global variables, it makes it hard to follow. When should I use these globals and when not. I think it's better if it's handled automatically somehow.

Which globals are you referring to?

**Comment by youknowriad** [comment]
Sorry I meant the environment variables

**Comment by aduth** [comment]
I don't know that we should need to use the environment variables for any other package. It's more exclusively affecting Babel plugins where we expect to use the (built) plugin in the building of other packages. The other option is to rewrite the plugin so it doesn't require a build step (ES5).

**Comment by youknowriad** [comment]
Ok I see, I think I'm fine with what we have. We can see if it's a lot of struggle over time and if we'd have to differentiate between "node/babel packages" and browser ones.

File: packages/babel-plugin-import-jsx-pragma/src/index.js (hunk 9 above)
Comment (aduth) [reply packages/babel-plugin-import-jsx-pragma/src/index.js]: Updated in 67653d0. Makes things simpler overall too.
